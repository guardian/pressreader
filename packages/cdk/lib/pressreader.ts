import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import { GuScheduledLambda } from '@guardian/cdk/lib/patterns/scheduled-lambda';
import type { App } from 'aws-cdk-lib';
import { Duration } from 'aws-cdk-lib';
import { Schedule } from 'aws-cdk-lib/aws-events';
import { Runtime } from 'aws-cdk-lib/aws-lambda';

export class PressReader extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);
		const pressReaderApp = 'pressreader';
		new GuScheduledLambda(this, `${pressReaderApp}-lambda`, {
			app: pressReaderApp,
			runtime: Runtime.NODEJS_18_X,
			memorySize: 512,
			handler: 'handler.main',
			fileName: `pressreader.zip`,
			monitoringConfiguration: {
				noMonitoring: true,
			},
			rules: [{ schedule: Schedule.cron({ minute: '0', hour: '1' }) }],
			timeout: Duration.seconds(300),
		});
	}
}
