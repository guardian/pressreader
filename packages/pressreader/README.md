# Pressreader

This is a scheduled lambda! 