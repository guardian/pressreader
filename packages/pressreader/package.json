{
  "name": "pressreader",
  "version": "1.0.0",
  "description": "",
  "devDependencies": {
    "@types/node": "18.13.0",
    "aws-sdk": "^2.1362.0"
  },
  "scripts": {
    "dev": "ts-node-dev --respawn src/handler.ts",
    "typecheck": "tsc -noEmit",
    "format": "prettier --write \"src/**/*.ts\"",
    "build": "tsc --noEmit & esbuild src/handler.ts --bundle --minify --outfile=dist/handler.js --external:aws-sdk --platform=node",
    "test": "jest --detectOpenHandles --config ../../jest.config.js --selectProjects pressreader"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/guardian/newsroom-resilience-platform.git"
  },
  "author": "",
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/guardian/newsroom-resilience-platform/issues"
  },
  "homepage": "https://github.com/guardian/newsroom-resilience-platform#readme",
  "dependencies": {
    "ts-node-dev": "^2.0.0"
  }
}
