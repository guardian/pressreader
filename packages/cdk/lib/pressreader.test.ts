import { App, Duration } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Schedule } from 'aws-cdk-lib/aws-events';
import { PressReader } from './pressreader';

describe('The PressReader stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new PressReader(app, 'PressReader', {
			stack: 'print-production',
			stage: 'TEST',
			lambdaConfigs: [
				{ editionKey: 'AUS', s3PrefixPath: ['data', 'AUS'] },
				{ editionKey: 'US', s3PrefixPath: ['data', 'US'] },
				{
					editionKey: 'AUS',
					s3PrefixPath: [],
					bucketName: 'press-reader-aus-configs',
				},
				{
					editionKey: 'US',
					s3PrefixPath: [],
					bucketName: 'press-reader-us-configs',
				},
			],
			schedule: Schedule.rate(Duration.minutes(15)),
			domainName: 'pressreader.test.dev-gutools.co.uk',
		});
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
