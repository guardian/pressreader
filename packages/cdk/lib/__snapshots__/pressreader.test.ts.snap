// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The PressReader stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuScheduledLambda",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "pressreaderlambdaB481BEF0": {
      "DependsOn": [
        "pressreaderlambdaServiceRoleDefaultPolicy7471EACC",
        "pressreaderlambdaServiceRole68181491",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "print-production/TEST/pressreader/pressreader.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "pressreader",
            "STACK": "print-production",
            "STAGE": "TEST",
          },
        },
        "Handler": "handler.main",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "pressreaderlambdaServiceRole68181491",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "pressreader",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/pressreader",
          },
          {
            "Key": "Stack",
            "Value": "print-production",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "pressreaderlambdaServiceRole68181491": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "pressreader",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/pressreader",
          },
          {
            "Key": "Stack",
            "Value": "print-production",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "pressreaderlambdaServiceRoleDefaultPolicy7471EACC": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/print-production/TEST/pressreader/pressreader.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/print-production/pressreader",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/print-production/pressreader/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "pressreaderlambdaServiceRoleDefaultPolicy7471EACC",
        "Roles": [
          {
            "Ref": "pressreaderlambdaServiceRole68181491",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "pressreaderlambdapressreaderlambdacron0106BFB30AD": {
      "Properties": {
        "ScheduleExpression": "cron(0 1 * * ? *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "pressreaderlambdaB481BEF0",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "pressreaderlambdapressreaderlambdacron010AllowEventRulePressReaderpressreaderlambda88EE711E420D72B5": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "pressreaderlambdaB481BEF0",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "pressreaderlambdapressreaderlambdacron0106BFB30AD",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;
